ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/fold_device_operation.cpp

```diff
@@ -0,0 +1,118 @@
+    }
+}
+
+MorehFoldOperation::program_factory_t MorehFoldOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return ProgramFactory{};
+};
+
+void MorehFoldOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+void MorehFoldOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehFoldOperation::shape_return_value_t MorehFoldOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.input.get_logical_shape();
+    auto input_tensor_rank = tensor_args.input.get_logical_shape().rank();
+    uint32_t kernel_size_product = operation_attributes.kernel_size[0] * operation_attributes.kernel_size[1];
+    if (input_tensor_rank == 3) {
+        uint32_t N = input_tensor_shape[0];
+        uint32_t C = input_tensor_shape[1] / kernel_size_product;
+        auto output_shape = ttnn::SimpleShape(
+            std::vector<uint32_t>{N, C, operation_attributes.output_size[0], operation_attributes.output_size[1]});
+        return output_shape;
+    }
+    // If input_tensor_rank == 2
+    uint32_t C = input_tensor_shape[0] / kernel_size_product;
+    auto output_shape = ttnn::SimpleShape(
+        std::vector<uint32_t>{C, operation_attributes.output_size[0], operation_attributes.output_size[1]});
+    return output_shape;
+};
+
+MorehFoldOperation::tensor_return_value_t MorehFoldOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    if (tensor_args.output.has_value()) {
+        return tensor_args.output.value();
+    }
+
+    const auto& output_shape = compute_output_shapes(operation_attributes, tensor_args);
+
+    return create_device_tensor(
+        output_shape,
+        tensor_args.input.get_dtype(),
+        tensor_args.input.get_layout(),
+        tensor_args.input.device(),
+        operation_attributes.memory_config);
+}
+
+std::tuple<MorehFoldOperation::operation_attributes_t, MorehFoldOperation::tensor_args_t> MorehFoldOperation::invoke(
+    const Tensor& input,
+    const std::optional<Tensor>& output,
+    const std::vector<uint32_t> output_size,
+    const std::vector<uint32_t> kernel_size,
+    const std::vector<uint32_t> dilation,
+    const std::vector<uint32_t> padding,
+    const std::vector<uint32_t> stride,
+    const std::optional<MemoryConfig>& memory_config) {
+    return {
+        operation_attributes_t{
+            output_size, kernel_size, dilation, padding, stride, memory_config.value_or(input.memory_config())},
+
+        tensor_args_t{input, output}};
```

Comment:
```suggestion
        operation_attributes_t{
            output_size, kernel_size, dilation, padding, stride, memory_config.value_or(input.memory_config())},
        tensor_args_t{input, output}};
```
It looks like `clang-format` hasn't been run, given the presence of this extra whitespace.

Reply:
I just ran clang format manually and there is still extra whitespace there. I will remove it then.

Reply:
resolved in latest commit.